### PR TITLE
Prevent time out when main thread is blocked

### PR DESCRIPTION
**Pull request by Johni0702**
Ticket: #411 
This prevents the player from getting kicked during or after an expensive operation in the main thread by moving the necessary checks into another thread which will never be blocked by any expensive operations.
This thread would also be the place where one would put a potential future watchdog system for the main thread.
Testing can be done with any expensive main-thread-locking plugin such as WorldEdit or ScriptCraft (`/js java.lang.Thread.sleep(30000)`) by blocking the main thread for more than 30 seconds.

---

_Edited by turt2live_

**Related Links:**
- Ticket: #411

### CODE REVIEW

**Comment by gdude2002** [comment]
Looks good. http://bamboo.gserv.me/browse/GLOW-SRV-PR415-83

File: src/main/java/net/glowstone/net/GlowSession.java [hunk 1]
```diff
@@ -107,6 +107,12 @@
     private int writeTimeoutCounter = 0;
 
     /**
+     * Whether to kick the player the next time its updated. This is set to {@code true}
```
Comment (mastercoms) [inline src/main/java/net/glowstone/net/GlowSession.java]: Grammar Nazi alert: it's, not its

**Comment by turt2live** [comment]
@Johni0702 this currently has a major bug where the console can (and will) be spammed with "Timed out" kick messages if the player is sent packets after they have been kicked once.

Scenario: I started 150 bots on a server and did a 17.3 million block change in WorldEdit. While WorldEdit was processing the change, I disconnected 40 bots using a hard disconnect (therefore leaving the server connection open). Once WorldEdit stopped saturating the main thread, the block updates were distributed and all 40 of those clients got a few million messages in the console each. After the kick message spam stopped, I did a smaller change (5000 blocks) and once again got the same spam message.

Here's an excerpt from the log on a smaller test to verify that this PR was at fault:

```
14:24:34 [INFO] testBot_113 kicked: Timed out
14:24:34 [INFO] testBot_44 kicked: Timed out
14:24:34 [INFO] testBot_116 kicked: Timed out
14:24:34 [INFO] testBot_113 kicked: Timed out
14:24:34 [INFO] testBot_44 kicked: Timed out
14:24:34 [INFO] testBot_116 kicked: Timed out
14:24:34 [INFO] testBot_113 kicked: Timed out
14:24:34 [INFO] testBot_44 kicked: Timed out
14:24:34 [INFO] testBot_116 kicked: Timed out
14:24:34 [INFO] testBot_113 kicked: Timed out
14:24:34 [INFO] testBot_44 kicked: Timed out
14:24:34 [INFO] testBot_116 kicked: Timed out
14:24:34 [INFO] testBot_113 kicked: Timed out
14:24:34 [INFO] testBot_44 kicked: Timed out
14:24:34 [INFO] testBot_116 kicked: Timed out
14:24:34 [INFO] testBot_113 kicked: Timed out
14:24:34 [INFO] testBot_44 kicked: Timed out
```

**Comment by turt2live** [comment]
@Johni0702 New issue: Sending ~9 million block changes to 153 players caused the clients to timeout from the server (normal), but the server fell into a full saturation of all resources well after the block changes were sent (I was peeking at the server while doing this). Looks like there's an infinite loop somewhere.

**Comment by Johni0702** [comment]
@turt2live I cannot seem to find any reason for the full saturation of all resources. Neither can I reproduce it as I do not have the required hardware to do so. Would you mind to test it again at some point and get me a thread dump of when it happens? When doing so, please also note down after how many seconds stuff happens (WE operation duration, first timeouts, seconds after WE operation till the thread dump, etc.)  as that might help figuring out this problem.

**Comment by turt2live** [comment]
I'm putting this back into `Needs Testing` so I can attempt to get those thread dumps.

**Comment by dequis** [comment]
Will probably be replaced by #189

**Comment by turt2live** [comment]
I'm assigning this to @SpaceManiac to review simply because I have little knowledge about netty. Ideally a comparison between this and #189 should be done in the meantime ( @coelho ?)

**Comment by coelho** [comment]
@turt2live My PR completely obsoletes the old system that this PR relies on.

**Comment by Johni0702** [comment]
I'd personally also prefer @coelho's PR as it's far more nettyish and less complicated for doing the same job. If his PR doesn't have any mandatory issues feel free to close this one, it's not even worth looking into compared to coelho's.

**Comment by turt2live** [comment]
Closing in favour of #189.

**Pull request closed without merging**